File: ex3/cpp/handle_input.cpp
#include <iostream>
#include <future>
#include <thread>
#include <chrono>
#include <wiringPi.h>

int main()
{
    const int button = 5; // < LED: Physical pin 18, BCM GPIO24, and WiringPi pin 5

    bool gpio_state = true, gpio_state_last = true;
    unsigned int cnt = 0;

    std::chrono::milliseconds timeout(100);
    std::future<int> async_getchar = std::async(std::getchar);

    wiringPiSetup();

    pinMode(button, INPUT);

    std::cout << "Press ENTER to exit." << std::endl;

    while(1)
    {
        gpio_state = (digitalRead(button) == HIGH);

        if (!gpio_state && gpio_state_last)
        {
            cnt++;
            std::cout << "Push-button counter: " << cnt <<std::endl;
        }

        gpio_state_last = gpio_state;

        if(async_getchar.wait_for(timeout) == std::future_status::ready)
        {
            async_getchar.get();
            break;
        }
    }

    return 0;
}

File: ex3/python/handle_input.py
#!/usr/bin/python3

import time
import sys
import select

try:
    import RPi.GPIO as GPIO
except RuntimeError:
    print("Error importing RPi.GPIO! This is probably because you need superuser privileges. You can achieve this by using 'sudo' to run your script.")

timeout = 0.1
buttonState = True
buttonStateLast = True
cnt = 0
i = ''

# Pin Definitions:
buttonPin = 18 # < Push-button: Physical pin 18, BCM GPIO24

# Pin Setup:
GPIO.setmode(GPIO.BOARD)
GPIO.setup(buttonPin, GPIO.IN) # Button pin set as input

print("Press ENTER to exit.")

while not i:
    # Waiting for I/O completion
    i, o, e = select.select([sys.stdin], [], [], timeout)

    if i:
        sys.stdin.readline()
        GPIO.cleanup() # cleanup all GPIO
        exit()

    buttonState = (GPIO.input(buttonPin) == GPIO.HIGH)

    if (buttonState is False) and (buttonStateLast is True):
        cnt += 1
        print("Push-button counter: ", cnt)

    buttonStateLast = buttonState
